### PR TITLE
NEW AdminController as superclass for /admin/* routed controllers

### PR DESCRIPTION
**Pull request by GuySartorelli**
There are two commits here:

1. Mostly-unrelated refactor to use config instead of runtime code to remove the `CMSProfileController` menu item
2. Add the new `AdminController` class and wire it up

I've also added a bunch of strong typing to the new class and to `AdminRouteController`.
I have _not_ added strong typing to `LeftAndMain` though that may happen as part of a follow-up card in this epic.
I have _not_ added strong typing to methods which would required extensive trickle-down effects (e.g. `init()` and `canView()`).

## Issue
- https://github.com/silverstripe/silverstripe-admin/issues/1761

### CODE REVIEW

File: _config.php [hunk 1]
```diff
@@ -31,5 +31,3 @@
     ]);
 // enable ability to insert anchors
 $editorConfig->insertButtonsAfter('sslink', 'anchor');
-
-CMSMenu::remove_menu_class(CMSProfileController::class);
```
Comment (ChloeSartorelli) [inline _config.php:35]: That's handled by config in the class itself now.

File: code/AdminController.php
Comment (ChloeSartorelli) [inline code/AdminController.php:1]: Basically everything in here was just lifted straight from `LeftAndMain` - though some small code quality improvements were made here-and-there.

File: code/AdminController.php [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+
+    /**
+     * Get list of required permissions for accessing this controller.
+     * If false, no permission is required.
+     */
+    public static function getRequiredPermissions(): array|string|false
+    {
+        if (static::class === AdminController::class) {
+            throw new BadMethodCallException('getRequiredPermissions should be called on a subclass');
+        }
```
Comment (ChloeSartorelli) [inline code/AdminController.php:73]: Just to avoid it being called here, since this is explicitly abstract there's no permission for this class (or reason to want one)

File: code/AdminController.php [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+
+        return true;
+    }
+
+    public function Link($action = null)
+    {
+        // LeftAndMain methods have a top-level uri access
+        if (static::class === LeftAndMain::class) {
+            $segment = '';
```
Comment (ChloeSartorelli) [inline code/AdminController.php:133]: A follow-up card will make `LeftAndMain` abstract after removing the methods on it that are used from singleton - but for now this has to stay.

File: code/AdminController.php [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+                $newResponse->addHeader($k, $v);
+            }
+            $newResponse->setIsFinished(true);
+            $this->setResponse($newResponse);
+            // Actual response will be re-requested by client
+            return $newResponse;
+        } else {
+            return parent::redirect($url, $code);
+        }
+    }
```
Comment (ChloeSartorelli) [inline code/AdminController.php:184]: I'm not 100% sure exactly what all of the ramifications of this are - but since it's dealing explicitly with AJAX requests and that's mostly what this method is for, I decided to lift this out of LeftAndMain and into this class.

File: code/AdminController.php [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+        // and use in routing definitions.
+        $url = trim($this->Link(), '/');
+        $clientConfig = [
+            'name' => $name,
+            'url' => $url,
+            'reactRoutePath' => preg_replace('/^' . preg_quote(AdminRootController::admin_url(), '/') . '/', '', $url),
+        ];
+        $this->extend('updateClientConfig', $clientConfig);
+        return $clientConfig;
+    }
```
Comment (ChloeSartorelli) [inline code/AdminController.php:203]: This is lifted over here so that we can inject links for these controllers into the config that JS gets a hold of.

File: code/AdminController.php [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+        // set language
+        $member = Security::getCurrentUser();
+        if (!empty($member->Locale)) {
+            i18n::set_locale($member->Locale);
+        }
+
+        // Don't allow access if the request hasn't finished being handled and the user can't access this controller
+        if (!$this->canView() && !$this->getResponse()->isFinished()) {
+            // Allow subclasses and extensions to redirect somewhere more appropriate
+            $this->invokeWithExtensions('onInitPermissionFailure');
```
Comment (ChloeSartorelli) [inline code/AdminController.php:223]: This lets LeftAndMain do some redirecting it was doing before without making the method abstract (and therefore required on classes that don't need it) and without doing a `method_exists()` check which always feels like the wrong move.

File: code/AdminController.php [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+                ),
+                'alreadyLoggedIn' => _t(
+                    __CLASS__ . '.PERMALREADY',
+                    "I'm sorry, but you can't access that part of the CMS."
+                ),
+                'logInAgain' => _t(
+                    __CLASS__ . '.PERMAGAIN',
+                    "You have been logged out of the CMS."
+                ),
+            ];
```
Comment (ChloeSartorelli) [inline code/AdminController.php:248]: Shortened these messages to suit the fact that most of the time these controllers will be handling AJAX requests and therefore there won't be a menu "below" like the old messages suggest.

I considered finding a way to use the longer messages in LeftAndMain specifically, but the context of "there's a form below" is unnecessary because... well, there's a form below. You can see it (or tab to it if you're on a screen reader)

File: code/AdminRootController.php [hunk 8]
```diff
@@ -10,24 +10,41 @@
...
     /**
      * Fallback admin URL in case this cannot be infered from Director.rules
+     */
+    private static string $url_base = 'admin';
+
+    /**
+     * The LeftAndMain child that will be used as the initial panel to display if none is selected (i.e. if you
+     * visit /admin)
+     */
+    private static string $default_panel = SecurityAdmin::class;
```
Comment (ChloeSartorelli) [inline code/AdminRootController.php:30]: These are just moved from below per our coding standards (i.e. properties before methods)

File: code/AdminRootController.php [hunk 9]
```diff
@@ -36,55 +53,31 @@ public static function get_admin_route()
...
      */
-    public static function rules()
+    public static function rules(): array
     {
         if (AdminRootController::$adminRules === null) {
             AdminRootController::$adminRules = [];
 
             // Map over the array calling add_rule_for_controller on each
-            $classes = CMSMenu::get_cms_classes(null, true, CMSMenu::URL_PRIORITY);
+            $classes = CMSMenu::get_cms_classes(AdminController::class, true, CMSMenu::URL_PRIORITY);
```
Comment (ChloeSartorelli) [inline code/AdminRootController.php:71]: `null` defaults to `LeftAndMain` - we want the new controller as the base class.

File: code/AdminRootController.php [hunk 10]
```diff
@@ -118,7 +111,7 @@ public function handleRequest(HTTPRequest $request): HTTPResponse
         // Otherwise
         $rules = AdminRootController::rules();
         foreach ($rules as $pattern => $controller) {
-            if (($arguments = $request->match($pattern, true)) !== false) {
+            if ($request->match($pattern, true) !== false) {
```
Comment (ChloeSartorelli) [inline code/AdminRootController.php:114]: Unrelated refactoring since that variable is never used.

File: code/CMSProfileController.php [hunk 11]
```diff
@@ -22,6 +22,8 @@ class CMSProfileController extends LeftAndMain
 
     private static $tree_class = Member::class;
 
+    private static $ignore_menuitem = true;
```
Comment (ChloeSartorelli) [inline code/CMSProfileController.php:25]: This is the config that lets us remove the code from `_config.php`

File: code/LeftAndMain.php [hunk 12]
```diff
@@ -281,16 +244,35 @@ class LeftAndMain extends Controller implements PermissionProvider
...
+     */
+    private static $application_link = '//www.silverstripe.org/';
+
+    /**
+     * The application name
+     *
+     * @config
+     * @var string
+     */
+    private static $application_name = 'Silverstripe';
```
Comment (ChloeSartorelli) [inline code/LeftAndMain.php]: Moved from below, along with `searchFilterCache` as well. These were below some methods.

File: code/LeftAndMain.php [hunk 13]
```diff
@@ -312,7 +301,7 @@ class LeftAndMain extends Controller implements PermissionProvider
     public function getCombinedClientConfig()
     {
         $combinedClientConfig = ['sections' => []];
-        $cmsClassNames = CMSMenu::get_cms_classes(LeftAndMain::class, true, CMSMenu::URL_PRIORITY);
+        $cmsClassNames = CMSMenu::get_cms_classes(AdminController::class, true, CMSMenu::URL_PRIORITY);
```
Comment (ChloeSartorelli) [inline code/LeftAndMain.php:304]: The config itself can be on any `AdminController` now.

File: code/SudoModeController.php [hunk 14]
```diff
@@ -91,7 +89,7 @@ public function activate(HTTPRequest $request): HTTPResponse
             return $this->jsonResponse([
                 'result' => false,
                 'message' => _t(__CLASS__ . '.INVALID', 'Incorrect password'),
-            ]);
+            ], 400);
```
Comment (ChloeSartorelli) [inline code/SudoModeController.php]: Unrelated refactoring - this is an error, so should have an error response code!  There's no context in the original PR that created this class for why it was a 200. It looks like it was an oversight.

File: tests/php/AdminControllerTest.php [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+        foreach ($allowed as $allowedClass) {
+            $this->assertTrue($allowedClass::singleton()->canView(), 'Tried to access ' . $allowedClass);
+        }
+        $disallowed = array_diff($allControllers, $allowed);
+        foreach ($disallowed as $disallowedClass) {
+            $this->assertFalse($disallowedClass::singleton()->canView(), 'Tried to access ' . $disallowedClass);
+        }
+    }
+
+    public static function provideJsonSuccess(): array
```
Comment (ChloeSartorelli) [inline tests/php/AdminControllerTest.php:98]: This and everything below it was pulled from `LeftAndMainTest`.

File: tests/php/LeftAndMainTest.php [hunk 16]
```diff
@@ -80,8 +74,7 @@ protected function tearDown(): void
 
     public function testExtraCssAndJavascript()
     {
-        $admin = $this->objFromFixture(Member::class, 'admin');
-        $this->logInAs($admin);
+        $this->logInWithPermission('ADMIN');
```
Comment (ChloeSartorelli) [inline tests/php/LeftAndMainTest.php:77]: The admin fixture and the `logInWithPermission('ADMIN')` user have the same permissions, so no need for the fixture.

File: tests/php/LeftAndMainTest.php [hunk 17]
```diff
@@ -135,65 +128,6 @@ public function testLeftAndMainSubclasses()
         $this->assertMatchesRegularExpression('/<body[^>]*>/i', $response->getBody(), "$link should contain <body> tag");
     }
 
-    public function testCanView()
```
Comment (ChloeSartorelli) [inline tests/php/LeftAndMainTest.php:138]: This was testing the wrong thing - it was trying to test `canView()` checks by seeing what's in the menu. But the menu actually _shouldn't_ contain `CMSProfileController` so this test was passing when it should have failed.

Moved canView check to the new test class, and now actually testing canView instead of the menu.

The menu itself is already tested in this class.

File: code/SudoModeController.php (hunk 14 above)
Comment (emteknetnz) [reply code/SudoModeController.php]: This should be 401, context for this is getting password wrong when trying to activate sudo mode.

https://stackoverflow.com/a/32752617

File: code/AdminRootController.php [hunk 18]
```diff
@@ -10,24 +10,41 @@
...
+
+    /**
+     * The LeftAndMain child that will be used as the initial panel to display if none is selected (i.e. if you
+     * visit /admin)
+     */
+    private static string $default_panel = SecurityAdmin::class;
+
+    /**
+     * Holds an array of url_pattern => controller k/v pairs, the same as Director::rules. However this is built
+     * dynamically from introspecting on all the classes that derive from LeftAndMain.
```
Comment (emteknetnz) [inline code/AdminRootController.php]:
```suggestion
     * dynamically from introspecting on all the classes that derive from AdminController.
```

File: code/AdminController.php [hunk 19]
```diff
@@ -0,0 +1,359 @@
...
+            Security::permissionFailure($this, $messageSet);
+            return;
+        }
+
+        // Don't continue if there's already been a redirection request.
+        if ($this->getResponse()->isRedirect()) {
+            return;
+        }
+
+        $this->extend('onInit');
```
Comment (emteknetnz) [inline code/AdminController.php:260]: Weird place for an 'onInit' hook - seems like it should be at the bottom?  Any reason why it's here? Does the implementation in LeftAndMain not work right if this hook is at the bottom?

File: code/LeftAndMain.php [hunk 20]
```diff
@@ -281,16 +244,35 @@ class LeftAndMain extends Controller implements PermissionProvider
...
+     * Set this to `false` via yml config if you do not want to show any help links
      *
      * @config
-     * @var string|null
+     * @var array
      */
-    private static $section_name = null;
+    private static $help_links = [
+        'CMS User help' => 'https://userhelp.silverstripe.org/en/5',
+        'Developer docs' => 'https://docs.silverstripe.org/en/5/',
```
Comment (emteknetnz) [inline code/LeftAndMain.php]:
```suggestion
        'CMS User help' => 'https://userhelp.silverstripe.org/en',
        'Developer docs' => 'https://docs.silverstripe.org/en/',
```

Comment (emteknetnz) [reply code/LeftAndMain.php]: Or /6/

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: code/SudoModeController.php (hunk 14 above)
Comment (ChloeSartorelli) [reply code/SudoModeController.php]: I hate HTTP response codes lol.
Updated to 401

File: code/AdminRootController.php (hunk 18 above)
Comment (ChloeSartorelli) [reply code/AdminRootController.php]: Done

File: code/AdminController.php (hunk 19 above)
Comment (ChloeSartorelli) [reply code/AdminController.php:260]: I'm just keeping it in the same logic place it was with `LeftAndMain`. I don't know if it would break anything if we move it, and I don't particularly want to find out. Chances are if it breaks anything to move it, we won't find out for a year or so.
I'm _guessing_ it's intentionally placed and that it's specifically not called `onAfterInit()` for that reason, though I can't confirm that.

Safer to just leave it where it is. But I can move it if you feel strongly.

File: code/LeftAndMain.php (hunk 20 above)
Comment (ChloeSartorelli) [reply code/LeftAndMain.php]: We'll need to update a bunch of references like this, we should probably just do them all together. But I'll make this change to avoid ping pong.
Going with an explicit `6` so people don't get docs for `7` in their `6` sites when we get to that point.

File: code/AdminController.php (hunk 19 above)
Comment (emteknetnz) [reply code/AdminController.php:260]: Oh right, hook was called 'init' in CMS 5, sorry didn't see that it was existing. Leave as is

**Review by emteknetnz — APPROVED** [review]
